.buildkite/generator.jl

```diff
@@ -0,0 +1,230 @@
+    else
+        Second(0)
+    end
+    delay > Second(0) && (msg *= " (" * string(Dates.canonicalize(round(delay, Second))) * " ago)")
+    @info(msg)
+    annotate(msg, style="info", context="pkg")
+    tolerance = Hour(1)
+    if delay > tolerance
+        @warn "The PkgServer registry is older than $(tolerance)"
+        annotate("The PkgServer registry is older than $(tolerance)", style = "warning", context="pkg")
+    end
+end
+
+# Get the directories holding changed files
+# 1. All changed files
+# 2. Only files in directories
+# 3. dirname
+# 4. Unique the directories
+PROJECTS = readlines(
+    pipeline(`git diff-tree --no-commit-id --name-only -r HEAD $(COMPARE_AGAINST)`,
+             `grep -E ".+/.+"`))
+PROJECTS = map(dirname, PROJECTS) |> unique!
+
+# If there are scary projects we need to exclude, we list them here. (Used to contain `LLVM`)
+EXCLUDED_NAMES = String[]
+
+# This is the dynamic mapping we're going to build up, if it's empty we don't do anything
+PROJECTS_ACCEPTED = Set{String}()
+for PROJECT in PROJECTS
+    println("Considering ", PROJECT)
+
+    # Only accept things that contain a `build_tarballs.jl`
+    while !ispath(joinpath(PROJECT, "build_tarballs.jl")) && !isempty(PROJECT)
+        println(" --> $(PROJECT) does not contain a build_tarballs.jl, moving up a directory")
+        PROJECT = dirname(PROJECT)
+    end
+
+    if isempty(PROJECT)
+        println(" --> Skipping as we could not find a build_tarballs.jl")
+        continue
+    end
+
+    # Ignore RootFS stuff, we'll do that manually
+    if startswith(PROJECT, "0_RootFS/")
+        println(" --> Skipping as it's within 0_RootFS/")
+        continue
+    end
+
+    NAME = basename(PROJECT)
+    # Ignore stuff in our excluded projects
+    if NAME in EXCLUDED_NAMES
+        println(" --> Skipping as it's excluded")
+        continue
+    end
+
+    # Otherwise, emit a build with `PROJECT` set to `${PROJECT}`
+    println(" --> Accepted!")
+    push!(PROJECTS_ACCEPTED, PROJECT)
+end
+
+if isempty(PROJECTS_ACCEPTED)
+    annotate("No projects accepted", style="success")
+    exit()
+end
+
+if length(PROJECTS_ACCEPTED) > 20
+    annotate("Too many projects", style="error")
+    exit(1)
+end
+
+YGGDRASIL_BASE = dirname(@__DIR__)
+julia(args) = `$(Base.julia_cmd()) --project=$(YGGDRASIL_BASE)/.ci $args`
+
+# Next, we're going to ensure that our BB is up to date and precompiled
+julia(`-e "import Pkg; Pkg.instantiate(); Pkg.precompile()"`) |> exec
+
+# We're going to snarf out the BB and BBB tree hashes and combine them to be used later in our build cache
+BB_HASH = julia(```
+    -e "using Pkg, SHA; \
+        gethash(uuid) = collect(Pkg.Types.Context().env.manifest[Pkg.Types.UUID(uuid)].tree_hash.bytes); \
+        print(bytes2hex(sha256(vcat( \
+            gethash(\"7f725544-6523-48cd-82d1-3fa08ff4056e\"), \
+            gethash(\"12aac903-9f7c-5d81-afc2-d9565ea332ae\"), \
+        ))));"
+```) |> readchomp
```

Comment:
Also here you can run the command locally without shelling out, but you may need to add `SHA` to the environment.

Reply:
I think we already have SHA.

Reply:
Right! https://github.com/JuliaPackaging/Yggdrasil/blob/9f9b73258dc3b347ee08b8c764b63c01cc462b44/.buildkite/Project.toml#L5

Reply:
Ok, reading what we're doing here more carefully, it's probably right we should run it in the `/.ci` environment more specifically, but I wonder if we should eventually merge the `.buildkite` and `.ci` environments.